### PR TITLE
feat(jenkinscontroller,azure-vm) JCasc revamp for initscript (Java Home, inbound, path)

### PR DESCRIPTION
**Pull request by dduportal**
Subject triggered by https://github.com/jenkins-infra/helpdesk/issues/3641, as a fixup of https://github.com/jenkins-infra/jenkins-infra/pull/2925 .

Switching from SSH launcher to inbound launcher changes the agent process environment (not a `login` session anymore but a Systemd service which does NOT source files such as `/etc/environment`).

Besides, since https://github.com/jenkins-infra/jenkins-infra/pull/2896, the Azure VM init script was duplicated between trusted.ci and azure.ci (and not yet enabled on cert.ci).

This PR introduces the following changes to the controller setups to solve these issues (some changes looks unrelated but are all required to avoid breaking existing setups):

- Initscript are now part of the Puppet template (instead of freely defined in hieradata) `clouds.yaml.erb` in the `jenkinscontroller` profile to ensure the same logic is used across all controllers.  
  - Enables datadog agent with any kind of launcher (SSH or inbound) with a requirement for a dtadog API key (added on cert.ci, rotated on ci.j and trusted.ci.j)
  - Support of both Linux and Windows
  - Inbound agent is installed as a service (systemd on Linux, WS on Windows)
  - Environment variables `ARTIFACT_CACHING_PROXY`, `JAVA_HOME` and `PATH` are properly defined (fix for https://github.com/jenkins-infra/helpdesk/issues/3641)

- Default JDK for agent is set to JDK11 **explictly** unless overriden (example: webbuilder container)
- Default agent working dir is changed to `/home/jenkins/agent` on Linux and `C:\Jenkins\agent` on Windows

### Improvements for subsequent PRs:

- Pass JVM options on Windows (already done on Linux) inbound azure VM agents
- Control environment on Windows in the same way as on Linux (at least `ARTIFACT_CACHING_PROXY`, `JAVA_HOME` and `Path`)
- Implement websocket flag passing in the init script for inbound Azure VM (both Linux/Windows)
- Implement direct TCP flag passing in the init script for inbound Azure VM (both Linux/Windows) for private network setups such as trusted.ci.j)
- Implement Kubernetes Clouds compliance with Windows

### CODE REVIEW

File: dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb [hunk 1]
```diff
@@ -11,13 +11,123 @@ jenkins:
...
+        <%- if agent['os'].to_s == 'windows' -%>
+        executeInitScriptAsRoot: false
+        # Escape '$' in JCasc with '^' - ref. https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/features/secrets.adoc#passing-secrets-through-variables
+        initScript: |
+          # Setup datadog
+          (Get-Content C:\ProgramData\Datadog\datadog.yaml -Raw) -Replace 'api_key:', 'api_key: <%= @jcasc['datadog_api_key'] %>' | Set-Content C:\ProgramData\Datadog\datadog.yaml
+          & "$env:ProgramFiles\Datadog\Datadog Agent\bin\agent.exe" restart
+            <%- if agent['launcher'] && agent['launcher'] == "inbound" -%>
+          # Setup inbound agent
+          $jenkinsserverurl = $args[0] # Always ends witgh a '/'
```
Comment (hervelemeur) [inline dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb]:
```suggestion
          $jenkinsServerUrl = $args[0] # Always ends with a '/'
```
Typo & camelCase

File: dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb [hunk 2]
```diff
@@ -11,13 +11,123 @@ jenkins:
...
+          $vmname = $args[1]
+          $secret = $args[2]
+          $jenkinsWorkDir = '<%= agent['agentDir'] ? agent['agentDir'] : @jcasc['agents_setup'][agent['os'].to_s]['agentDir'] %>'
+          # Download the service wrapper
+          $agentJar =  '{0}jnlpJars/agent.jar' -f $jenkinsserverurl
```
Comment (hervelemeur) [inline dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb]:
```suggestion
          $agentJar =  '{0}jnlpJars/agent.jar' -f $jenkinsServerUrl
```

File: dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb [hunk 3]
```diff
@@ -11,13 +11,123 @@ jenkins:
...
+            chmod 770 /var/log/datadog
+            systemctl daemon-reload
+            systemctl enable datadog-agent.service
+            systemctl start datadog-agent.service
+          ) |& tee /var/log/agent-init-datadog.log
+          <%- if agent['launcher'] && agent['launcher'] == "inbound" -%>
+          # Setup Jenkins Agent Service
+          (
+            # Argument provided by the Azure-VM plugin
+            export JENKINS_URL="^${1}" # Always ends witgh a '/'
```
Comment (hervelemeur) [inline dist/profile/templates/jenkinscontroller/casc/clouds.yaml.erb]:
```suggestion
            export JENKINS_URL="^${1}" # Always ends with a '/'
```

**Review by hervelemeur — APPROVED** [review]
Great job!

Small nitpicks

**Review by timja — APPROVED** [review]
personally I would extract out to a shell script file or powershell file to avoid all the escaping and have syntax highlighting etc but this should work

**Comment by dduportal** [comment]
- Thanks for the review. I've added more camel casing (I admit I do not care at all but better have an homogeneous script, easier to maintain with your suggestions)
- Tested on ci.jenkins.io, trusted.ci and cert.ci manually. There are still issues on trusted.ci and cert.ci that will be adressed on a subsequent PR (puppet agent disabled on both until then)

**Comment by dduportal** [comment]
> personally I would extract out to a shell script file or powershell file to avoid all the escaping and have syntax highlighting etc but this should work

Alas you need to generate 2 different scripts then: one for SSH and one for inbound. They have a code part in common (datadog) but also divergences (setting up env with SSH, setting up agent service with its own env impl in inbound): resolving this with puppet is easier.

**Pull request merged**